text/0000-conditional-compilation-checking.md

```diff
@@ -0,0 +1,486 @@
+
+```rust
+#[cfg(feature = "widnows")]
+fn do_windows_thing() { /* ... */ }
+```
+
+The developer intended to test for the feature named `windows`. This could easily have been detected
+by `rustc` if it had known the set of all valid `feature` flags, not only the ones currently
+enabled.
+
+This RFC proposes adding new command-line options to `rustc`, which will allow Cargo (and other
+build tools) to inform `rustc` of the set of valid feature flags. Using conditions that are not
+valid will cause a diagnostic warning. This feature is opt-in, for backwards compatibility;
+if no valid configuration options are presented to `rustc` then no warnings are generated.
+
+# Motivation
+
+* Stronger assurance that large code bases are correct.
+* Protect against typos, bad merges, etc.
+* Detect dead code, typically caused by feature flags that have been removed from a crate's
+  manifest, but which still have `#[cfg(...)]` attributes that mention those features.
+
+# Guide-level explanation
+
+## Background
+
+Rust programs can use conditional compilation in order to modify programs based on the features a
+user has selected, the target CPU architecture, the target OS, or other parameters under control
+of the user. Rust programs may use conditional compilation in these ways:
+
+* By applying the `#[cfg(c)]` attribute to language elements, where `c` is a condition.
+* By applying the `#[cfg_attr(c, attr)]` attribute to language elements, where `c` is a conditional
+  and `attr` is an attribute to apply.
+* By using the `cfg!(c)` built-in macro, where `c` is a condition. The compiler replaces the macro
+  call with a `true` or `false` literal.
+
+A _condition_ can take one of two forms:
+
+* A single identifier, such as `#[cfg(test)]` or `#[cfg(linux)]`. These are Boolean conditions;
+  they are either enabled or disabled.
+* A condition may test whether a given value is present in a named list of values. For example,
+  `#[cfg(feature = "lighting")]` tests whether the `lighting` feature is enabled. Note that a given
+  condition name may have any number of enabled values; for example, it is legal to invoke
+  `rustc --cfg feature="lighting" --cfg feature="bump_maps"`.
+
+## Checking conditions names
+
+`rustc` can optionally verify that all condition names used in source code are valid. "Valid" is
+distinct from "enabled". A _valid_ condition is one that is allowed to appear in source code; the
+condition may be enabled or disabled, but it is still valid. An _enabled_ condition is one which has
+been specified with a `--cfg foo` or `--cfg 'foo="value"'` option.
+
+For example, `rustc` can detect this bug, where the `test` condition is misspelled as `tset`:
+
+```rust
+if cfg!(tset) {    // uh oh, should have been 'test'
+   ...
+}
+```
+
+To catch this error, we give `rustc` the set of valid condition names:
+
+```bash
+rustc --cfg 'valid(name1, name2, ..., nameN)' ...
+```
+
+The `--cfg 'valid(...)'` option specifies the set of valid condition names.

```

Comment:
A possible solution to the other issues I raised is to also allow an attribute (maybe `#![valid(name1, name2, ...)]`). That way, you can say yourself the features you expect. Then the only `valid` option set by cargo will be for features, and rustc will know that cfgs other than features should not give warnings.

Reply:
> The feature proposed here feels very much desireable to me and actually I'm surprised that `rustc` doesn't already do this.
> A warning for a wrong `cfg` falls in the caregory of "Thank you compiler, you just safed me from a weird bug." :)
> 
> But the CLI syntax should be simplified:
> 
> ```shell
> rustc --cfg 'valid(name1, name2, ..., nameN)' ...
> rustc --cfg 'valid_values(c, "value1", "value2", ... "valueN")' ...
> ```
> 
> * There will surely be an environment that messes up the quotes or parentheses.
> * The format is inconsistent with other rustc arguments, which are like `--arg=val1,val2,val3`.
> * These are not really `--cfg` arguments, so they should have their own `--whatever` flags.
> 
> Alternatives that spring to my mind:
> 
> ```shell
> rustc --valid_cfgs=name1,name2,nameN
> # Not shure how the handle this "arg with a list" situation
> rustc --valid_values_for=cond1=value1,value2,valueN
> # Special-casing features might be worth it if it's the only thing that will be implemented
> rustc --valid_features=feature1,feature2,featureN 
> ```

I'm certainly open to changing how the options are represented. I wrote the spec this way for two reasons. First, it didn't require adding any new command-line options. Second, the code that handles the existing `--cfg` options already parses them as metadata, so enabling a new syntactic form for that metadata seemed easy. It also removed the need to handle passing more `--foo` options down into that part of the compiler.

Agreed that dealing with quoting is ugly. The single-quotes in this are mainly just there for pedantic illustrative purposes; they just show clearly which string goes into which element of `argv`. `rustc` already requires double-quotes when passing `--cfg feature="foo"`, so tools already have to deal with quoting these strings. The only that my proposal adds is the `(` and `)`, for grouping the parts of the metadata.

Still, I'm open to changing these.  Perhaps `--cfg_valid foo,bar,etc` for the condition names, and `--cfg_valid_values feature="foo","bar","etc"` for checking values?

Reply:
> Still, I'm open to changing these. Perhaps --cfg_valid foo,bar,etc for the condition names, and --cfg_valid_values feature="foo","bar","etc" for checking values?

That sounds good to me :) and it's also more consistent with how the other cargo arguments work.

Reply:
Yeah, I think I'm convinced that having separate args is better. That way it's clear that the meaning of `--cfg` doesn't change. It also helps for build tools (and humans) that have to parse rustc command lines. That's not a common need, but it does crop up sometimes with tools like compilers.

I'll edit the doc.  `--cfg valid(foo,bar)` will become `--cfg-valid=foo,bar`, and `--cfg valid_values(key,"value")` will become `--cfg-valid-values 'key = "value"'`.